kong/plugins/aws-lambda/v4.lua

```diff
@@ -0,0 +1,228 @@
+    else
+      segments[#segments+1] = urldecode(segment):gsub("[^%w%-%._~]", percent_encode)
+    end
+  end
+  local len = #segments
+  if len == 0 then return "/" end
+  -- If there was a slash on the end, keep it there.
+  if path:sub(-1, -1) == "/" then
+    len = len + 1
+    segments[len] = ""
+  end
+  segments[0] = ""
+  segments = table.concat(segments, "/", 0, len)
+  return segments
+end
+
+local function canonicalise_query_string(query)
+  local q = {}
+  for key, val in query:gmatch("([^&=]+)=?([^&]*)") do
+    key = urldecode(key):gsub("[^%w%-%._~]", percent_encode)
+    val = urldecode(val):gsub("[^%w%-%._~]", percent_encode)
+    q[#q+1] = key .. "=" .. val
+  end
+  table.sort(q)
+  return table.concat(q, "&")
+end
+
+local function derive_signing_key(kSecret, date, region, service)
+  local kDate = hmac("AWS4" .. kSecret, date)
+  local kRegion = hmac(kDate, region)
+  local kService = hmac(kRegion, service)
+  local kSigning = hmac(kService, "aws4_request")
+  return kSigning
+end
+
+local function prepare_awsv4_request(tbl)
+  local domain = tbl.domain or "amazonaws.com"
+  local region = tbl.region
+  local service = tbl.service
+  local request_method = tbl.method
+  local canonicalURI = tbl.canonicalURI
+  local path = tbl.path
+  if path and not canonicalURI then
+    canonicalURI = canonicalise_path(path)
+  elseif canonicalURI == nil or canonicalURI == "" then
+    canonicalURI = "/"
+  end
+  local canonical_querystring = tbl.canonical_querystring
+  local query = tbl.query
+  if query and not canonical_querystring then
+    canonical_querystring = canonicalise_query_string(query)
+  end
+  local req_headers = tbl.headers or {}
+  local req_payload = tbl.body
+  local access_key = tbl.access_key
+  local signing_key = tbl.signing_key
+  local secret_key
+  if not signing_key then
+    secret_key = tbl.secret_key
+    if secret_key == nil then
+      return nil, "either 'signing_key' or 'secret_key' must be provided"
+    end
+  end
+  local timestamp = tbl.timestamp or os.time()
+  local tls = tbl.tls
+  if tls == nil then tls = true end
+  local port = tbl.port or (tls and 443 or 80)
+  local req_date = os.date("!%Y%m%dT%H%M%SZ", timestamp)
+  local date = os.date("!%Y%m%d", timestamp)
+
+  local host = service .. "." .. region .. "." .. domain
+  local host_header do -- If the "standard" port is not in use, the port should be added to the Host header
+    local with_port
+    if tls then
+      with_port = port ~= 443
+    else
+      with_port = port ~= 80
+    end
+    if with_port then
+      host_header = string.format("%s:%d", host, port)
+    else
+      host_header = host
+    end
+  end
+
+  local headers = {
+    ["X-Amz-Date"] = req_date;
+    Host = host_header;
+  }
+  local add_auth_header = true
+  for k, v in pairs(req_headers) do
+    k = k:gsub("%f[^%z-]%w", string.upper) -- convert to standard header title case
+    if k == "Authorization" then
+      add_auth_header = false
```

Comment:
Would this break if you were using oauth2 protected apis with the use of the Authorization header? Causing the lambda call to not function properly?